src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
 
             stateMachineType = new IteratorStateMachine(slotAllocatorOpt, compilationState, method, methodOrdinal, isEnumerable, elementType);
             compilationState.ModuleBuilderOpt.CompilationState.SetStateMachineType(method, stateMachineType);
-            return new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics).Rewrite();
+            var rewriter = new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics);
+            if (rewriter.EnsureAllSymbolsAndSignature())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        // returns true if any type we depend on is not present
+        protected bool EnsureAllSymbolsAndSignature()
+        {
+            bool hasErrors = false;
+
+            EnsureSpecialType(SpecialType.System_Int32, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator<T>
+            EnsureSpecialPropertyAsMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current,
+                                            SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
+
+            // IEnumerator
+            EnsureSpecialProperty(SpecialMember.System_Collections_IEnumerator__Current, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__MoveNext, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__Reset, ref hasErrors);
```

Comment:
It looks like each `EnsureSpecial*` method calls `F.SpecialType(type)`. Consider getting the type once for each special type.